esi_leap/objects/flocx_market_client.py

```diff
@@ -63,3 +64,64 @@ def send_offer(self, data):
                 r.status_code,
                 r.text)
         return r.status_code
+
+    def get_contracts(self):
+        url = CONF.flocx_market.endpoint_override + \
+        '/offer_contract_relationship?status=unretrieved'
+        r = self.get(url)
+        if r.status_code != 200:
+            LOG.warning(
+                "Failed to retrieve contracts from flocx-market. Got HTTP %s: %s",
+                r.status_code,
+                r.text)
+        if r.content is not None:
+            contracts = []
+            contract_offer_table = {}
+            contract_offer_list = json.loads(r.content)
+            # create contract offers hash table
+            # key: contract_id, value: (offer_id, offer_contract_id) list
+            for c_o in contract_offer_list:
+                if not contract_offer_table.get(str(c_o["contract_id"])):
+                    o_c_r_tuple = (str(c_o["marketplace_offer_id"]), str(c_o["offer_contract_relationship_id"]))
+                    contract_offer_table[str(c_o["contract_id"])] = [o_c_r_tuple]
+                    contract_offer_table[str(c_o["contract_id"])].append(o_c_r_tuple)
+
+            # retrieve contracts and offers data from flocx-market iteratively
+            for contract_id, o_c_rs in contract_offer_table.items():
+                contract_data = {}
+                url1 = CONF.flocx_market.endpoint_override + '/contract/' + contract_id
+                r_contract = self.get(url1)
+                c = json.loads(r_contract.content)
+                contract_data["contract_id"] = str(c["contract_id"])
+                # contract_data["time_created"] = str(c["time_created"])
+                contract_data["start_time"] = str(c["start_time"])
+                contract_data["end_time"] = str(c["end_time"])
+                contract_data["status"] = str(c["status"])
+                contract_data["cost"] = str(c["cost"])
+                contract_data["bid_id"] = str(c["bid_id"])
+                contract_data["offers"] = []
+                contract_data["marketplace_offer_id"] = []
+                contract_data["o_c_rs"] = []
+                for o_c_r in o_c_rs:
+                    url2 = CONF.flocx_market.endpoint_override + '/offer/' + o_c_r[0]
+                    r_offer = self.get(url2)
+                    o = json.loads(r_offer.content)
+                    contract_data["offers"].append(str(o["provider_offer_id"]))
+                    contract_data["marketplace_offer_id"].append(str(o["marketplace_offer_id"]))
+                    contract_data["o_c_rs"].append(str(o_c_r[1]))
+                contracts.append(contract_data)
+            return contracts
+        else:
+            return []
+
```

Comment:
You're trying to be efficient here with the hash table, to avoid repeated queries. However the code is a bit hard to read and debug, and in the long-term we'll have a single query to deal with all of this. I suggest doing this the straightforward way: after you retrieve all the ocrs, iterate over the ocrs, and for each ocr query the contract from the marketplace and then the bid from the marketplace.

Reply:
Okay, I've changed the code a bit. Now provider queries marketplace very straightforward: get the offer_contract_pair, make get request to contract, offer, bid and create the contract. I was thinking to form the contract data in the structure that a single query should provide, like, using a single query provider can get the contracts with multiple offers, and it needs to split the contracts to create new ones. But for now provider skip the 'split contract' step.